### PR TITLE
SF-1170: Add support for all(), allOff(), and subscribeWith() methods

### PR DESCRIPTION
**Pull request by jrmykolyn**

### CODE REVIEW

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+    it('should reset the event counters for a given collection', () => {
+      const events = ['a', 'b'];
+      const key = 'a:b';
+      const cb = spy();
+
+      emitter.all(events, cb);
+      events.forEach(ev => emitter.emit(ev, null));
+
+      expect(Object.keys(emitter._barriers[key].events).map(k => emitter._barriers[key].events[k])).to.eql([0, 0]);
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: Use `Object.values`. The order in which `Object.keys` and `Object.values` return their contents is unspecified, but in this case, it doesn't matter since their values should all be 0.

Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: As discussed, `Object.values` was introduce by the ES017 spec, which is not targeted by our `tsconfig` file. Leaving as-is.

**Comment by jrmykolyn** [comment]
**RE: `Object.values`:**

As discussed, `Object.values` was introduce by the ES017 spec, which is not targeted by our `tsconfig` file.

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  emit(event: string, ...args: any[]): boolean {
+    const result = super.emit(event, ...args);
+    const keys = this._lookups[event] || [];
+
+    keys.forEach((key) => {
+      const inc = this._barriers[key].events[event] + 1;
+      this._barriers[key].events = { ...this._barriers[key].events, [event]: inc };
+
+      const shouldInvoke = Object.keys(this._barriers[key].events).every(ev => this._barriers[key].events[ev] >= 1);
```
[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: As discussed, Object.values was introduce by the ES017 spec, which is not targeted by our tsconfig file. Leaving as-is.

[the remark this answers is not in the document]
Comment (Vic-Dev) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: I think we've had problems with using Object.values in the past @jmbtutor (ie?)

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    return this;
+  }
+
+  generateKey(events: string[]) {
+    return events.slice(0).sort().join(':');
+  }
+}
+
+export interface E {
+  [key: string]: number;
```
[the remark this answers is not in the document]
Comment (Vic-Dev) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: What's the benefit of doing this? https://stackoverflow.com/questions/51936369/what-is-the-record-type-in-typescript
Tldr; I think it's less clear

File: packages/@storefront/flux-capacitor/src/core/emitter.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmbtutor) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: It somehow slipped my mind that `Object.values` was introduced after ES2015. Yes, we should leave this as-is.

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  all(events: string[], cb: () => void, context = this) {
+    const key = this.generateKey(events);
+
+    this._barriers[key] = {
+      events: events.reduce((acc, ev) => ({ ...acc, [ev]: 0 }), {}),
+      cbs: this._barriers[key] ? [...this._barriers[key].cbs, { cb, context }] : [{ cb, context }],
+    };
+
+    this._lookups = events.reduce((acc, ev) => ({...acc, [ev]: this._lookups[ev] ? [...this._lookups[ev], key] : [key] }), this._lookups);
```
[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: Updated implementation to use `Object.assign()` instead of spread. Use of `reduce` left as-is.

File: packages/@storefront/flux-capacitor/CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased] [minor]
+### Added
+- SF-1170: Add support for `all()` and `allOff()` methods.
```
Comment (Vic-Dev) [inline packages/@storefront/flux-capacitor/CHANGELOG.md]: Maybe add `emitter` or `event listener` support... (there are many parts of flux that change, eg store, etc, might be useful to point out specifically what this is for)

File: packages/@storefront/core/tsconfig.esnext.json [hunk 6]
```diff
@@ -6,7 +6,9 @@
       "DOM"
     ],
     "types": [
-      "node"
+      "node",
```
Comment (Vic-Dev) [inline packages/@storefront/core/tsconfig.esnext.json]: You shouldn't have to change the `esnext` tsconfigs :s

File: packages/@storefront/core/tsconfig.json [hunk 7]
```diff
@@ -7,7 +7,9 @@
       "DOM"
     ],
     "types": [
-      "node"
+      "node",
```
Comment (Vic-Dev) [inline packages/@storefront/core/tsconfig.json]: @jmbtutor should we do this as a separate PR?

Comment (jmbtutor) [reply packages/@storefront/core/tsconfig.json]: Yes.

Comment (Vic-Dev) [reply packages/@storefront/core/tsconfig.json]: 👍 Jesse can you make a separate PR for the tsconfig stuff? I see the ticket is in "ready for review" but I don't see a PR for it

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    this._barriers[key] = {
+      events: events.reduce((acc, ev) => ({ ...acc, [ev]: 0 }), {}),
+      callbacks: this._barriers[key]
+        ? [...this._barriers[key].callbacks, { callback, context }]
+        : [{ callback, context }],
+    };
+
+    // tslint:disable-next-line max-line-length
+    this._lookups = events.reduce((acc, ev) => Object.assign(acc, { [ev]: this._lookups[ev] ? [...this._lookups[ev], key] : [key] }), this._lookups);
```
[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: @jmbtutor asked for `Object.assign()`. Consensus?

**Comment by jrmykolyn** [comment]
Added `Do Not Merge` label. To be removed when #10 flows into `master`.

File: packages/@storefront/flux-capacitor/src/core/emitter.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jmbtutor) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: We do this `Object.assign` pattern elsewhere in the code, and using `Object.assign` means that the amount of garbage that you create with the intermediary objects is just O(n) instead of O(n<sup>2</sup>). Ideally, we create no garbage by doing a for-each and assigning directly.

On that note, re-assigning to `this._lookups` is redundant because you're assigning the object to itself, since `this._lookups` is used as the initial accumulator.

File: packages/@storefront/core/tsconfig.json (hunk 7 above)
Comment (jrmykolyn) [reply packages/@storefront/core/tsconfig.json]: @Vic-Dev Yup, submitted as #10.

File: packages/@storefront/core/tsconfig.esnext.json (hunk 6 above)
Comment (jrmykolyn) [reply packages/@storefront/core/tsconfig.esnext.json]: Change introduce as part of #10. Marking as resolved.

File: packages/@storefront/core/test/unit/core/tag/index.ts [hunk 9]
```diff
@@ -171,4 +170,70 @@ suite('Tag', ({ expect, spy, stub }) => {
...
+         [[events[2]], cb2],
+         [[events[3], events[4], events[5]], cb3]
+       ];
+       tag.flux = { allOff: spy() };
+
+       tag._removeLookups();
+
+       expect(tag.flux.allOff.args).to.eql(tag._lookups);
+    });
+  });
```
Comment (jmbtutor) [inline packages/@storefront/core/test/unit/core/tag/index.ts]: Is it worth testing the case for non-disjoint events?

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
+import { EventEmitter } from 'eventemitter3';
+
+class Emitter extends EventEmitter {
+  _barriers: Barriers;
+  _lookups: Lookups;
+
+  constructor() {
+    super();
+    this._barriers = {};
+    this._lookups = {};
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/src/core/emitter.ts]: TypeScript allows you to put properties right on the class:

```
class Emitter extends EventEmitter {
  _barriers: Barriers = {};
  _lookups: Lookups = {};

  // ...
}
```

File: packages/@storefront/flux-capacitor/src/core/emitter.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Vic-Dev) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: https://github.com/Microsoft/TypeScript/issues/13689#issuecomment-275285593

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+    return this;
+  }
+
+  generateKey(events: string[]) {
+    return events.slice(0).sort().join(':');
+  }
+}
+
+export interface Events {
```
[the remark this answers is not in the document]
Comment (Vic-Dev) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: O right mb

File: packages/@storefront/flux-capacitor/src/core/emitter.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Vic-Dev) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: mb didn't realize he asked you to switch to it. It's not worth arguing over.

[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: Leaving as-is, marking as resolved.

File: packages/@storefront/flux-capacitor/src/core/emitter.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: Leaving as-is, marking as resolved.

File: packages/@storefront/flux-capacitor/src/core/emitter.ts (hunk 10 above)
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: Will apply.

**Comment by jrmykolyn** [comment]
Removing `Do Not Merge` label (#10 has flowed into `master`).

File: packages/@storefront/flux-capacitor/src/core/emitter.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: As per Slack discussion: leaving as-as; marking as resolved.

File: packages/@storefront/core/test/unit/core/tag/index.ts (hunk 9 above)
Comment (jrmykolyn) [reply packages/@storefront/core/test/unit/core/tag/index.ts]: Discussed IRL: not relevant for proposed implementation. Marking as resolved.

File: packages/@storefront/core/CHANGELOG.md [hunk 12]
```diff
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased] [minor]
+### Added
+- SF-1170: Add support for `subscribeWith()` method.
+  - Allows components to invoke callbacks after each event in a given collection has been emitted at least once.
```
[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/core/CHANGELOG.md]: Implemented.

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+        0,
+        1,
+        null,
+        undefined,
+        {},
+        () => null,
+        'Hello, world!',
+      ];
+
+      vals.forEach((val) => expect(() => emitter.all(val , () => null)).to.throw(error));
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: Pull the `forEach` up and run an `it` for each case. Put the value in the description of the `it`.

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+      callback1 = spy();
+      callback2 = spy();
+    });
+
+    it('should return the emitter' , () => {
+      expect(emitter.all(['a', 'b', 'c'], () => null)).to.eq(emitter);
+    });
+
+    it('should throw an error when `events` is not an array', () => {
+      const error = '`events` is not an array.';
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: The error message is insignificant; that is, the test should still pass since changing the string doesn't change the behaviour. Given that, I don't think it should be included in the test.

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+
+  describe('allOff', () => {
+    let events1;
+    let events2;
+    let key1;
+    let key2;
+    let callback1;
+    let callback2;
+
+    before(() => {
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: The tests are not isolated with this as a `before`. This should be a `beforeEach`.

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+        key1 = 'a\nb';
+        callback1 = spy();
+        callback2 = spy();
+    });
+
+    it('should update the _barriers counters if they exist', () => {
+      emitter.all(events1, () => null);
+      emitter.emit('a', null);
+
+      expect(emitter._barriers[key1].events.a).to.equal(1);
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: `.to.eq`

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+    });
+  });
+
+  describe('generateKey', () => {
+    it('should generate a sorted barrier key', () => {
+      const events = ['c', 'd', 'a', 'z'];
+
+      const key = emitter.generateKey(events);
+
+      expect(key).to.equal('a\nc\nd\nz');
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: `.to.eq`

File: packages/@storefront/flux-capacitor/tsconfig.esnext.json [hunk 18]
```diff
@@ -5,7 +5,9 @@
       "ES2016.Array.Include"
     ],
     "types": [
-      "node"
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/tsconfig.esnext.json]: Revert file.

File: packages/@storefront/core/tsconfig.esnext.json (hunk 6 above)
Comment (jmbtutor) [inline packages/@storefront/core/tsconfig.esnext.json]: Revert file.

**Comment by Vic-Dev** [comment]
Did you re-demo after all these changes?

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts (hunk 13 above)
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: Implemented.

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts (hunk 15 above)
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: Implemented.

File: packages/@storefront/flux-capacitor/test/unit/core/emitter.ts (hunk 14 above)
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/test/unit/core/emitter.ts]: Implemented.

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+  _lookups: Lookups = {};
+
+  emit(event: string, ...args: any[]): boolean {
+    const result = super.emit(event, ...args);
+    const keys = this._lookups[event] || [];
+
+    keys.forEach((key) => {
+      this._barriers[key].events[event]++;
+
+      const shouldInvoke = Object.keys(this._barriers[key].events).every((ev) => this._barriers[key].events[ev] > 0);
```
[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: Implement a variation of this feedback.

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+      barrier.callbacks = barrier.callbacks.filter(({ callback: fn }) => fn !== callback);
+
+      const hasCallbacks = !!barrier.callbacks.length;
+
+      if (!hasCallbacks) {
+        delete this._barriers[key];
+      }
+
+      events.forEach((ev) => {
+        if (!hasCallbacks) {
```
[the remark this answers is not in the document]
Comment (jrmykolyn) [reply packages/@storefront/flux-capacitor/src/core/emitter.ts]: You're right, good call! Implemented.

File: packages/@storefront/flux-capacitor/src/core/emitter.ts [hunk 21]
```diff
@@ -35,13 +36,14 @@ class Emitter extends EventEmitter {
...
-        callbacks: [],
-      };
+      this._barriers[key] = [];
     }
 
-    this._barriers[key].callbacks.push({ callback, context });
+    this._barriers[key].push({
+      callback,
+      context,
+      events: events.reduce((acc, ev) => ({ ...acc, [ev]: 0 }), {}),
```
Comment (jmbtutor) [inline packages/@storefront/flux-capacitor/src/core/emitter.ts]: `Object.assign` for consistency.
